libraries/objectDefaultFiles/BabylonjsWorker.js

```diff
@@ -0,0 +1,389 @@
+        
+                        // singnal end of frame
+                        this.messageInterface.postMessage({
+                            workerId: this.workerId,
+                            isFrameEnd: true,
+                        });
+                        this.clientState = BabylonjsWorker.STATE_FRAME_DONE;
+                        break;
+                    default:
+                        console.error("wrong state for generating frames clientState: " + this.clientState);
+                }
+            } else if (message.name === "context_lost") {
+                switch (this.clientState) {
+                    case BabylonjsWorker.CONSTRUCTED:
+                    case BabylonjsWorker.STATE_BOOTSTRAP_DONE:
+                    case BabylonjsWorker.STATE_FRAME_DONE:
+                    case BabylonjsWorker.STATE_CONTEXT_RESTORED_DONE:
+                        this.clientState = BabylonjsWorker.STATE_CONTEXT_LOST;
+                        this.onContextLost();
+                        break;
+                    default:
+                        console.error("wrong state for lost context clientState: " + this.clientState);
+                        break;
+                }
+            } else if (message.name === "context_restored") {
+                if (this.clientState === BabylonjsWorker.STATE_CONTEXT_LOST) {
+                    this.clientState = BabylonjsWorker.STATE_CONTEXT_RESTORED;
+                    this.onContextRestored();        
+                    this.clientState = BabylonjsWorker.STATE_CONTEXT_RESTORED_DONE;
+                } else {
+                    console.error("wrong state to restore state clientState: " + this.clientState);
+                }
+            }
+        }
+    }
+
+    onContextLost() {
+        this.glCommandBufferContext.onContextLost();
+        this.fakeCanvas.webglcontextlost({preventDefault: () => {}});
+    }
+
+    onContextRestored() {
+        this.glCommandBufferContext.onContextRestored();
+
+        let restoreBuffer = this.commandBufferFactory.createAndActivate(true);
+        this.fakeCanvas.webglcontextrestored();
+        restoreBuffer.execute();
+        
+        // singnal end of frame
+        this.messageInterface.postMessage({
+            workerId: this.workerId,
+            isFrameEnd: true,
+        });
+    }
+
+    /**
+     * registers callbacks for the tool to use during initalisation
+     * @param {function(BABYLON.Scene):void} callback 
+     * @returns {BabylonjsWorker}
+     */
+    onSceneCreated(callback) {
+        this.onSceneCreatedCallbacks.push(callback);
+        return this;
+    }
+
+    /**
+     * registers a callback for the tool to use during rendering
+     * @param {function(number):void} callback 
+     * @returns {BabylonjsWorker}
+     */
+    onRender(callback) {
+        this.onRenderCallbacks.push(callback);
+        return this;
+    }
+
+    /**
+     * starts three.js engine with the fake webgl context
```

Comment:
nit: references three.js here and on line 245. I'm guessing the constraints (fake canvas, etc.) are the same but wanted to point it out